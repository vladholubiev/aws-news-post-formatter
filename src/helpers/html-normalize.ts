import {load} from 'cheerio/lib/slim';
import {minify} from 'html-minifier';

export function normalizeHTML(rawHTML: string): string {
  const $ = load(rawHTML, null, false);

  $('*').each((_, element) => {
    $(element).removeAttr('style');
    $(element).removeAttr('class');
  });

  $('style').remove();

  $('td').each((_, element) => {
    const elTagNames = $(element)
      .find('*')
      .map((_, el) => el.tagName)
      .get();
    const elText = $(element).text();
    console.log(elTagNames);

    if (elTagNames.length === 1 && elTagNames[0] === 'br') {
      $(element).html(elText);
    }
  });

  // set tables to full width
  $('table').each((_, element) => {
    $(element).css('width', '100%');
  });

  // remove bgcolor from <td> and <table>
  $('td,table').each((_, element) => {
    $(element).removeAttr('bgcolor');
  });

  // remove empty <p> and <a> blocks like "<p>&nbsp;</p>"
  $('p,a').each((_, element) => {
    if ($(element).text().trim() === '') {
      $(element).remove();
    }
  });

  // fix relative urls
  $('a').each(function () {
    const href = $(this).attr('href');

    if (href?.includes('/content/aws.amazon.com/live/en_US/')) {
      const urlPath = href.split('/content/aws.amazon.com/live/en_US')[1].split('.html')[0];
      $(this).attr('href', `https://aws.amazon.com${urlPath}`);
    }

    if (href?.includes('/content/aws.amazon.com/staging/en_US/')) {
      const urlPath = href.split('/content/aws.amazon.com/staging/en_US/')[1].split('.html')[0];
      $(this).attr('href', `https://aws.amazon.com${urlPath}`);
    }
  });

  // remove double <b> tags
  $('b').each((_, element) => {
    const parent = $(element).parent();

    if (parent.is('b')) {
      $(element).contents().unwrap();
    }
  });

  // replace <b> with <strong>
  $('b').each((_, element) => {
    $(element).replaceWith(`<strong>${$(element).html()}</strong>`);
  });

  const finalHTML = $.root().html() || '';

  return minify(finalHTML, {
    collapseBooleanAttributes: true,
    collapseWhitespace: true,
    decodeEntities: true,
    minifyCSS: true,
    removeAttributeQuotes: true,
    removeComments: true,
    removeEmptyAttributes: true,
    removeOptionalTags: true,
    removeRedundantAttributes: true,
  });
}
