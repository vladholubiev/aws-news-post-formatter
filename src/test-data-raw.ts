import rawDataJSON from './test-data-raw.test.json';

type RawData = {
  post_body: string;
  size: number;
};

export const RAW_DATA = rawDataJSON as RawData[];
