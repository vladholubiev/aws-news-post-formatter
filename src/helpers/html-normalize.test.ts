import {RAW_DATA} from '../test-data-raw';
import {normalizeHTML} from './html-normalize';

it('should format a table w/o unnecessary styles', () => {
  const rawHTML = RAW_DATA[0].post_body;
  const formattedHTML = normalizeHTML(rawHTML);

  expect(formattedHTML).toMatchSnapshot();
});

it('should normalize complex html', () => {
  const rawHTML = RAW_DATA[62].post_body;
  const formattedHTML = normalizeHTML(rawHTML);

  expect(formattedHTML).toMatchSnapshot();
});

it('should normalize complex table html', () => {
  const rawHTML = RAW_DATA[1].post_body;
  const formattedHTML = normalizeHTML(rawHTML);

  expect(formattedHTML).toMatchSnapshot();
});

it('should fix formatting of a bullet list', () => {
  const rawHTML = RAW_DATA[20].post_body;
  const formattedHTML = normalizeHTML(rawHTML);

  expect(formattedHTML).toMatchSnapshot();
});

it('should fix formatting of a bullet list with bullet list inside single tag', () => {
  const rawHTML = RAW_DATA[9].post_body;
  const formattedHTML = normalizeHTML(rawHTML);

  expect(formattedHTML).toMatchSnapshot();
});

it('should render a table w/o unnecessary styles', async () => {
  const rawHTML = RAW_DATA[0].post_body;
  const formattedHTML = normalizeHTML(rawHTML);

  await page.goto('about:blank');
  await page.setContent(formattedHTML);
  const image = await page.screenshot({type: 'png', fullPage: true});

  expect(image).toMatchImageSnapshot();
});
