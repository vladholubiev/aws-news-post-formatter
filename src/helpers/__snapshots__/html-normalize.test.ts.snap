// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fix formatting of a bullet list 1`] = `"<p><a href=https://aws.amazon.com/cloudformation>AWS CloudFormation</a> provides a common language for you to describe and provision all the infrastructure resources in your cloud environment. It allows you to use a simple text file to model and provision, in an automated and secure manner, all the resources needed for your applications across all regions and accounts. <p>The following resources were updated as part of this release: AWS::EC2::Host, AWS::EC2::Route, AWS::EC2::VPNConnection, AWS::ECS::Cluster, AWS::ECS::Service, AWS::ECS::TaskDefinition, AWS::EFS::MountTarget, AWS::ElasticLoadBalancingV2::ListenerRule, AWS::EMR::Cluster, AWS::KinesisFirehose::DeliveryStream, AWS::S3::Bucket.<p><a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-host.html>AWS::EC2::Host</a><br>Use the HostRecovery property to indicates whether to enable or disable host recovery for the Dedicated Host.<p><a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-route.html>AWS::EC2::Route</a><br>Use the TransitGatewayId property to specify the ID of a transit gateway.<p><a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-vpn-connection.html>AWS::EC2::VPNConnection</a><br>Use the TransitGatewayId property to specify the ID of the transit gateway associated with the VPN connection.<p>Use the VpnGatewayId property to specify the ID of the virtual private gateway at the AWS side of the VPN connection.<p><a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecs-cluster.html>AWS::ECS::Cluster</a><br>Use the Tags property to apply metadata to clusters to help you categorize and organize them.<p><a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecs-service.html>AWS::ECS::Service</a><br>Use the EnableECSManagedTags property to specify whether to enable Amazon ECS managed tags for the tasks within the service.<p>Use the PropagateTags property to specify whether to propagate the tags from the task definition or the service to the tasks in the service.<p>Use the Tags property to apply metadata to services to help you categorize and organize them.<p><a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ecs-taskdefinition.html>AWS::ECS::TaskDefinition</a><br>In the <a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-containerdefinitions.html>ContainerDefinition</a> property type, use the ResourceRequirementsproperty to specify the type and amount of a resource to assign to a container. The only supported resource is a GPU.<p>In the <a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-containerdefinitions.html>ContainerDefinition</a> property type, use the Secrets property to specify the secrets to pass to the container.<p>Use the Tags property to apply metadata to task definitions to help you categorize and organize them.<p><a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-efs-mounttarget.html>AWS::EFS::MountTarget</a><br>Use the IpAddress attribute to return the IPv4 address of the mount target.<p><a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-listenerrule.html>AWS::ElasticLoadBalancingV2::ListenerRule</a><p>In the <a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-elasticloadbalancingv2-listenerrule-conditions.html>RuleCondition</a> property type:<p>● Use the HostHeaderConfig property to specify information for a host header condition.<p>● Use the HttpHeaderConfig property to specify information for an HTTP header condition.<p>● Use the HttpRequestMethodConfig property to specify information for an HTTP method condition.<p>● Use the PathPatternConfig property to specify information for a path pattern condition.<p>● Use the QueryStringConfig property to specify information for a query string condition.<p>● Use the SourceIpConfig property to specify information for a source IP condition.<p><a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticmapreduce-cluster.html>AWS::EMR::Cluster</a><br>In the <a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-elasticmapreduce-cluster-jobflowinstancesconfig.html>JobFlowInstancesConfig</a> property type, use the Ec2SubnetIdsproperty to specify multiple EC2 subnet IDs.<p><a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-kinesisfirehose-deliverystream-extendeds3destinationconfiguration.html>AWS::KinesisFirehose::DeliveryStream</a><br>In the <a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-kinesisfirehose-deliverystream-extendeds3destinationconfiguration.html>ExtendedS3DestinationConfiguration</a> property type:<br><p>● Use the DataFormatConversionConfiguration property to specify the serializer, deserializer, and schema for converting data from the JSON format to the Parquet or ORC format before writing it to Amazon S3.<p>● Use the ErrorOutputPrefix property to specify a prefix that Kinesis Data Firehose evaluates and adds to failed records before writing them to S3.<p>●The Prefix property is no longer required.<p>In the <a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-kinesisfirehose-deliverystream-s3destinationconfiguration.html>S3DestinationConfiguration</a> property type, use the ErrorOutputPrefix property to specify a prefix that Kinesis Data Firehose evaluates and adds to failed records before writing them to S3.<p><a href=https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/Welcome.html>AWS::S3::Bucket<br></a>Use the ObjectLockConfiguration property to specify an object lock configuration for the specified bucket.<p>Use the ObjectLockEnabled property to specify whether this bucket has an object lock configuration enabled."`;

exports[`should fix formatting of a bullet list with bullet list inside single tag 1`] = `"<p>You can now provision the following AWS resources using CloudFormation.<p>• <strong>AWS::EMR::SecurityConfiguration</strong><p>o <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-emr-securityconfiguration.html>Create</a> an Amazon EMR security configuration, which is stored in the service and can be specified when a cluster is created.<p>CloudFormation has also updated support for the following resources:<p>• <strong>AWS::AutoScaling::LifecycleHook</strong><p>o The <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-as-lifecyclehook.html#cfn-as-lifecyclehook-notificationtargetarn>NotificationTargetARN</a> and <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-as-lifecyclehook.html#cfn-as-lifecyclehook-rolearn>RoleARN</a> properties are now optional for <a href=https://aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-as-lifecyclehook.html>AWS::AutoScaling::LifecycleHook</a>.<p>• <strong>AWS::CloudWatch::Alarm</strong><p>o <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cw-alarm.html>Create</a> a CloudWatch alarm using the <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cw-alarm.html#cfn-cloudwatch-alarms-evaluatelowsamplecountpercentile>EvaluateLowSampleCountPercentile</a>, <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cw-alarm.html#cfn-cloudwatch-alarms-extendedstatistic>ExtendedStatistic</a>, and <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cw-alarm.html#cfn-cloudwatch-alarms-treatmissingdata>TreatMissingData</a> properties.<p>• <strong>AWS::EC2::SpotFleet</strong><p>o Now CloudFormation supports mutable changes to Spot fleet properties. The following properties of the <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-spotfleet-spotfleetrequestconfigdata.html>SpotFleetRequestConfigData</a> property support Replacement updates:<p>• AllocationStrategy<br>• IamFleetRole<br>• LaunchSpecifications<br>• SpotPrice<br>• TerminateInstancesWithExpiration<br>• ValidFrom<br>• ValidUntil<p>o The following properties of the <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-spotfleet-spotfleetrequestconfigdata.html>SpotFleetRequestConfigData</a> property support No interruption updates:<p>• ExcessCapacityTerminationPolicy<br>• TargetCapacity<p>• <strong>AWS::EMR::InstanceGroupConfig</strong><p>o Use <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-emr-instancegroupconfig.html#cfn-elasticmapreduce-instancegroupconfig-autoscalingpolicy>AutoScalingPolicy</a> property to support Auto Scaling for Amazon EMR task instance groups.<p>• <strong>AWS::Events::Rule</strong><p>o <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-events-rule.html>Use</a> the <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-events-rule-target.html#cfn-events-rule-target-rolearn>RoleArn</a> property on the <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-events-rule-target.html>Target</a> data type to specify the IAM role. The RoleArn property is deprecated on the Rule resource.<p>•	<strong>AWS::Kinesis::Stream</strong><p>o	The <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-kinesis-stream.html#cfn-kinesis-stream-shardcount>ShardCount</a> property for <a href="http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-kinesis-stream.html?shortFooter=truel">AWS::Kinesis::Stream</a> now supports No interruption updates. Use the <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-kinesis-stream.html#cfn-kinesis-stream-retentionperiodhourst>RetentionPeriodHours</a> property to configure the number of hours for the data records stored in shards to remain accessible<p>•	<strong>AWS::Lambda::Function</strong><p>o	<a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html>Use</a> the <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-function.html#cfn-lambda-function-tracingconfig>TracingConfig</a> property to configure tracing settings for your AWS Lambda function. For more information about tracing Lambda functions, see <a href=http://docs.aws.amazon.com/lambda/latest/dg/lambda-x-ray.html#using-x-ray>Tracing Lambda-Based Applications with AWS X-Ray</a> in the AWS Lambda Developer Guide.<p>•	<strong>AWS::Redshift::Cluster</strong><p>o	Use the <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-redshift-cluster.html>Tags</a> property to specify tags to associate with the Amazon Redshift cluster resource.<p>•	<strong>AWS::Redshift::ClusterParameterGroup</strong><p>o	Use the <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-redshift-clusterparametergroup.html>Tags</a> property to specify tags to associate with the Amazon Redshift parameter group resource.<p>•	<strong>AWS::Redshift::ClusterSecurityGroup</strong><p>o	Use the <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-redshift-clustersecuritygroup.html>Tags</a> property to specify resource tags on an Amazon Redshift security group.<p>•	<strong>AWS::Redshift::ClusterSubnetGroup</strong><p>o	Use the <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/AWSCloudFormation/latest/UserGuide/aws-resource-redshift-clustersubnetgroup.html>Tags</a> property to specify resource tags on an Amazon Redshift subnet group.<p>•	<strong>AWS::RDS::DBCluster</strong><p>o	<a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rds-dbcluster.html>Added</a> the ReadEndpoint.Address attribute to the <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-getatt.html>Fn::GetAtt</a> intrinsic function.<p>•	<strong>AWS::S3::Bucket</strong><p>o	<a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-s3-bucket.html>Added</a> the Amazon Resource Name ( Arn) attribute to the <a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-getatt.html>Fn::GetAtt</a> intrinsic function.<p>Please visit our website for more information on AWS CloudFormation:<br>•	<a href=https://aws.amazon.com/cloudformation>Product Page</a><br>•	<a href=https://aws.amazon.com/documentation/cloudformation>Documentation</a><br>•	<a href=http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-supported-resources.html>Supported AWS Resources</a> "`;

exports[`should format a table w/o unnecessary styles 1`] = `"<p>Starting today, Amazon EC2 M5, M5a, M5d, R5, R5a, and R5d instances will be available in new 8xlarge and 16xlarge sizes. The specifications for the new sizes are in the tables below. With these new instance sizes, customers who are currently using either m4.10xlarge, m4.16xlarge, r4.8xlarge, or r4.16xlarge now have an easy upgrade path to the latest generation of instances.  <table cellspacing=0 cellpadding=0 border=1 style=width:100%><tr><td width=114 valign=top><p align=center><strong><span>Size</span></strong><td width=66 valign=top><p align=center><strong><span>vCPUs</span></strong><td width=72 valign=top><p align=center><strong><span>Memory</span></strong><p align=center><strong><span>(GiB)</span></strong><td width=149 valign=top><p align=center><strong><span>Storage (GB)</span></strong><td width=114 valign=top><p align=center><strong><span>EBS-Optimized Bandwidth (Gbps)</span></strong><td width=109 valign=top><p align=center><strong><span>Network Bandwidth (Gbps)</span></strong><tr><td width=114 valign=top><p><span>m5.8xlarge</span><td width=66 valign=top><p align=center><span>32</span><td width=72 valign=top><p align=center><span>128</span><td width=149 valign=top><p align=center><span>EBS Only</span><td width=114 valign=top><p align=center><span>5</span><td width=109 valign=top><p align=center><span>10</span><tr><td width=114 valign=top><p><span>m5a.8xlarge</span><td width=66 valign=top><p align=center><span>32</span><td width=72 valign=top><p align=center><span>128</span><td width=149 valign=top><p align=center><span>EBS Only</span><td width=114 valign=top><p align=center><span>3.5</span><td width=109 valign=top><p align=center><span>Up to 10</span><tr><td width=114 valign=top><p><span>m5d.8xlarge</span><td width=66 valign=top><p align=center><span>32</span><td width=72 valign=top><p align=center><span>128</span><td width=149 valign=top><p align=center><span>2 x 600 NVMe SSD</span><td width=114 valign=top><p align=center><span>5</span><td width=109 valign=top><p align=center><span>10</span><tr><td width=114 valign=top><p><span>m5.16xlarge</span><td width=66 valign=top><p align=center><span>64</span><td width=72 valign=top><p align=center><span>256</span><td width=149 valign=top><p align=center><span>EBS Only</span><td width=114 valign=top><p align=center><span>10</span><td width=109 valign=top><p align=center><span>20</span><tr><td width=114 valign=top><p><span>m5a.16xlarge</span><td width=66 valign=top><p align=center><span>64</span><td width=72 valign=top><p align=center><span>256</span><td width=149 valign=top><p align=center><span>EBS Only</span><td width=114 valign=top><p align=center><span>7</span><td width=109 valign=top><p align=center><span>12</span><tr><td width=114 valign=top><p><span>m5d.16xlarge</span><td width=66 valign=top><p align=center><span>64</span><td width=72 valign=top><p align=center><span>256</span><td width=149 valign=top><p align=center><span>4 x 600 NVMe SSD</span><td width=114 valign=top><p align=center><span>10</span><td width=109 valign=top><p align=center><span>20</span></table><p><a href=https://aws.amazon.com/ec2/instance-types/m5/ target=_blank>M5, M5a, and M5d instances</a> are general purpose instances and are ideal for business-critical applications, web and application servers, back-end servers for enterprise applications, gaming servers, caching fleets, and app development environments.<table cellspacing=0 cellpadding=0 border=1 style=width:100%><tr><td width=114 valign=top><p><strong><span>Size</span></strong><td width=66 valign=top><p><strong><span>vCPUs</span></strong><td width=72 valign=top><p><strong><span>Memory</span></strong><p><strong><span>(GiB)</span></strong><td width=149 valign=top><p><strong><span>Storage (GB)</span></strong><td width=114 valign=top><p><strong><span>EBS-Optimized Bandwidth (Gbps)</span></strong><td width=109 valign=top><p><strong><span>Network Bandwidth (Gbps)</span></strong><tr><td width=114 valign=top><p><span>r5.8xlarge</span><td width=66 valign=top><p><span>32</span><td width=72 valign=top><p><span>256</span><td width=149 valign=top><p><span>EBS Only</span><td width=114 valign=top><p><span>5</span><td width=109 valign=top><p><span>10</span><tr><td width=114 valign=top><p><span>r5a.8xlarge</span><td width=66 valign=top><p><span>32</span><td width=72 valign=top><p><span>256</span><td width=149 valign=top><p><span>EBS Only</span><td width=114 valign=top><p><span>3.5</span><td width=109 valign=top><p><span>Up to 10</span><tr><td width=114 valign=top><p><span>r5d.8xlarge</span><td width=66 valign=top><p><span>32</span><td width=72 valign=top><p><span>256</span><td width=149 valign=top><p><span>2 x 600 NVMe SSD</span><td width=114 valign=top><p><span>5</span><td width=109 valign=top><p><span>10</span><tr><td width=114 valign=top><p><span>r5.16xlarge</span><td width=66 valign=top><p><span>64</span><td width=72 valign=top><p><span>512</span><td width=149 valign=top><p><span>EBS Only</span><td width=114 valign=top><p><span>10</span><td width=109 valign=top><p><span>20</span><tr><td width=114 valign=top><p><span>r5a.16xlarge</span><td width=66 valign=top><p><span>64</span><td width=72 valign=top><p><span>512</span><td width=149 valign=top><p><span>EBS Only</span><td width=114 valign=top><p><span>7</span><td width=109 valign=top><p><span>12</span><tr><td width=114 valign=top><p><span>r5d.16xlarge</span><td width=66 valign=top><p><span>64</span><td width=72 valign=top><p><span>512</span><td width=149 valign=top><p><span>4 x 600 NVMe SSD</span><td width=114 valign=top><p><span>10</span><td width=109 valign=top><p><span>20</span></table><p><a href=https://aws.amazon.com/ec2/instance-types/r5/ target=_blank>R5, R5a, and R5d instances</a> are memory optimized instances and are ideal for high performance databases, distributed web scale in-memory caches, mid-size in-memory databases, real time big data analytics, and other enterprise applications.<p>M5, M5a, M5d, R5, R5a, and R5d instances are now available in 8 sizes, with 2, 4, 8, 16, 32, 48, 64, and 96 vCPUs and can be purchased as On-Demand, Reserved or Spot Instances. These instances are available in all the AWS Regions where the existing sizes are already available.<br><p><span>To get started, visit the </span><a href=https://console.aws.amazon.com/ target=_blank>AWS Management Console</a><span>, </span><a href=https://aws.amazon.com/cli/ target=_blank>AWS Command Line Interface (CLI)</a><span>, and </span><a href=http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/EC2.html target=_blank>AWS SDKs</a><span>. To learn more, visit the </span><a href=https://aws.amazon.com/ec2/instance-types/m5/ target=_blank>Amazon EC2 M5 instance page</a><span> or the </span><a href=https://aws.amazon.com/ec2/instance-types/r5/ target=_blank>Amazon EC2 R5 instance page</a><span>.</span>"`;

exports[`should normalize complex html 1`] = `"<p>AWS Launch Wizard now allows customers to deploy SAP workloads using Red Hat Enterprise Linux Version 8.1. <p>AWS Launch Wizard offers a guided way of sizing, configuring, and deploying AWS resources for SAP HANA and SAP HANA-based Netweaver systems with a purpose built, easy to use wizard.  <p>On-boarding a new operating system version for an SAP application involves close collaboration between SAP and Operating System (OS) teams, who analyze SAP notes, configure the system, and test extensively before standardizing on the new OS. AWS Launch Wizard eliminates this heavy lifting. You select the OS you want to use from a dropdown menu, and it automatically creates a pre-configured environment aligned with AWS and SAP best practices. <p>Configuring RHEL 8.1 for SAP workloads manually requires several additional steps compared to RHEL 7, including installation of basic OS packages, C-shell, and SAP HANA High Availability cluster configuration commands. With AWS Launch Wizard, these tasks are automated, simplifying configuration and reducing time to deployment.<br><p>The following table shows the operating systems currently supported for different SAP components by AWS Launch Wizard:<br><p>AWS Launch Wizard supports multiple deployment patterns, provides AWS resource recommendations, and is integrated with AWS compute, storage, management, logging, monitoring, and security services. It frees up employee/partner resources from undifferentiated infrastructure build activities so they can focus on differentiating business process design and implementation activities. <p>To learn more about using AWS Launch Wizard to accelerate your SAP deployments, visit the <a href=https://aws.amazon.com/launchwizard>Launch Wizard page</a>, <a href=https://aws.amazon.com/launchwizard/faq>FAQ</a>, and <a href=https://docs.aws.amazon.com/launchwizard/latest/userguide/launch-wizard-sap.html>overview documentation</a>.<table style=width:100%><tr><td><br><strong>Operating system version</strong><td><strong>Single-Instance deployment</strong><td><br><strong>ASCS</strong><td><br><strong>ERS</strong><td><br><strong>PAS</strong><td><br><strong>SAP HANA database</strong><td><br><strong>SAP HANA database in HA cluster</strong><tr><td>Red-Hat-Enterprise-Linux-8.1-For-SAP-HA-US-HVM<td>✓<td>✓<td>✓<td>✓<td>✓<td>✓<tr><td>SuSE-Linux-12-SP5-For-SAP-HVM<td>✓<td>✓<td>✓<td>✓<td>✓<td>✓<tr><td>SuSE-Linux-12-SP5-For-SAP-BYOS-HVM<td>✓<td>✓<td>✓<td>✓<td>✓<td>✓<tr><td>SuSE-Linux-15-SP1-HVM<td>✓<td> <td> <td>✓<td>✓<td> <tr><td>SuSE-Linux-15-SP1-For-SAP-HVM<td>✓<td>✓<td>✓<td>✓<td>✓<td>✓<tr><td>SuSE-Linux-15-SP1-For-SAP-BYOS-HVM<td>✓<td>✓<td>✓<td>✓<td>✓<td>✓<tr><td>Red-Hat-Enterprise-Linux-7.6-For-SAP-HA-US-HVM<td>✓<td>✓<td>✓<td>✓<td>✓<td>✓<tr><td>Red-Hat-Enterprise-Linux-8.1-For-SAP-HA-US-HVM<td> <td> <td> <td> <td> <td> <tr><td>SuSE-Linux-15-HVM<td>✓<td> <td> <td>✓<td>✓<td> <tr><td>SuSE-Linux-15-For-SAP-HVM<td>✓<td>✓<td>✓<td>✓<td>✓<td>✓<tr><td>SuSE-Linux-15-For-SAP-BYOS-HVM<td>✓<td>✓<td>✓<td>✓<td>✓<td>✓<tr><td>SuSE-Linux-12-SP4-HVM<td>✓<td> <td> <td>✓<td>✓<td> <tr><td>SuSE-Linux-12-SP4-For-SAP-HVM<td>✓<td>✓<td>✓<td>✓<td>✓<td>✓<tr><td>SuSE-Linux-12-SP4-For-SAP-BYOS-HVM<td>✓<td>✓<td>✓<td>✓<td>✓<td>✓<tr><td>SuSE-Linux-12-SP5-HVM<td>✓<td> <td> <td>✓<td>✓<td> </table>"`;
